### ISSUE TITLE
Front end changes for service type field (FL)

### ISSUE THREAD

**Issue by mmiddaugh** [department-of-veterans-affairs/va.gov-team#26852]
## Issue description
Two issues were discovered with the he service type field. 
1. After selecting facility type = "Community providers (in VA's network)", user could type free text into service type and execute search -> resulting in unsuccessful searches without clear guidance about the error or way to have a successful search
2. Screen reader users were not made aware of typeahead field -> causing a frustrating and inaccessible experience 

Although the usability and functionality of the service type box is quirky and unfriendly, we focused on the lowest hanging fruit improvements which can be made to immediately benefit users. A complete redesign of this field will be needed in the future.

## Tasks

- [ ] Implement design changes outlined in Nick's design consistent with [UX Pin](https://preview.uxpin.com/602167f0ca1beccf4bdf706e04b3a5bd34c78e94#/pages/139907964/documentation/no-panels?mode=i)  
- [ ] Implement intended screen reader/aria messaging 
- [ ] Implement error messaging notes

## Acceptance criteria

**Expected behavior for initial state**
- [ ] On input focus, with no text in the field, or fewer than two characters, a dropdown prompt appears: "Search for an available service"
- [ ] Screen reader announces prompt when it displays
- [ ] Input retains focus, prompt is not accessible via arrow keys

**Expected behavior if there is no service type match** 
- [ ] If no service type matches are found, a prompt is displayed: "We couldn't find that, please try another service"
- [ ] Screen reader announces prompt when it displays
- [ ] Input retains focus, prompt is not accessible via arrow keys

**Expected behavior if search is attempted with no valid service selection** 
- [ ] If search is attempted with no valid service selection, a prompt is displayed: "Please search for an available service." and no search is executed
- [ ] Current screen reader and focus management for this scenario is maintained

**Comment by mpelzsherman** [comment]
We're currently using Downshift v.1.22.5, which is very old.

If we would benefit greatly from upgrading, we should consider it, but keep in mind impact to other teams. (Probably a separate ticket to upgrade.)

**Comment by mpelzsherman** [comment]
Suggestion: break this ticket out into separate tickets:

1. Upgrade Downshift
2. Require valid service type
3. a11y improvements

**Comment by gescobardev** [comment]
Separated Downshift version upgrade issue here - https://github.com/department-of-veterans-affairs/va.gov-team/issues/28359

**Comment by tlei123** [comment]
@joshkimux , I just added a Regression AC above in the main-description.  FYI, although Search worked fine with mouse-nav after successful Service-type selection, the error-prompt seemed to pop-up after **Tab**ing to **Search** button and pressing **Enter**.   
 
![Screen Shot 2021-08-12 at 5 27 39 PM](https://user-images.githubusercontent.com/587583/129286741-597290c1-33ad-456f-b49a-b64b88736eaa.png)

@mpelzsherman , if this kybd-nav defect should be logged, pls tag Josh to open a proper A11y ticket.

\* I usually don't QA kybd-nav since that's under A11y-coverage, but occasionally do check that when I have the time, like just now.

**Comment by tlei123** [comment]
@mpelzsherman @gescobardev , mouse-nav QA on macOS Chrome & Firefox passed\*.  [See above Comment re. kybd-nav though].

\* RIs are inaccessible via Safari, BrowserStack mobile devices, or Windows browsers, due to SOCKS-proxy support-limitations.

**Comment by mpelzsherman** [comment]
Good catch, @tlei123 ! Here are the steps I had to take to reproduce the keyboard nav issue:

1. Tab to Facilty Type dropdown, select Community providers
2. Tab to Service Type. Type "Gen", select a service type
3. Shift-tab back to Facility Type
4. Tab back to Service Type. Error message is displayed.

This is an existing bug in production, but we should try to fix it in this ticket anyway.

**Comment by joshkimux** [comment]
# A11y feedback:

## Focus can get visibly lost for sighted keyboard users on longer lists
Type in "Psy" and use the down arrow key to navigate through the list. If the list is long enough and the user's screen does not have enough height, focus will become lost and out of sight.

https://user-images.githubusercontent.com/14154792/129373082-6815d04e-ca42-4d5c-b2df-35e23ab2b762.mov

**Suggestion:** Limiting the max height of the suggestions list before overflow to around 5-10 items (e.g. 25vh) could reduce the chances of this occuring.

### Before
<img width="473" alt="Screen Shot 2021-08-13 at 10 36 11 AM" src="https://user-images.githubusercontent.com/14154792/129373603-bda544a1-3b5d-4f92-9bd6-2abf85d57a07.png">

### After
<img width="472" alt="Screen Shot 2021-08-13 at 10 35 56 AM" src="https://user-images.githubusercontent.com/14154792/129373618-360ff7a4-a95d-460b-93a5-c7fb93b9b988.png">


## Some mobile users may not have access to a bluetooth keyboard
Instructions are still provided on mobile to use arrow keys to navigate through options. Folks using a mobile screen reader device may not have a bluetooth keyboard on hand to do so.

<img width="844" alt="Screen Shot 2021-08-13 at 10 24 12 AM" src="https://user-images.githubusercontent.com/14154792/129371990-0ea8d4c5-1dba-407a-9618-0765df5a43ed.png">

**Suggestion:** I would recommend reaching out to @AngelaFowler82 for confirmation on what the preferred screen reader experience should be here on mobile.

**Comment by mpelzsherman** [comment]
I like the suggestion to limit the height of the list. If we can do this easily, I think we should go ahead and add that requirement to this ticket.

The screen reader message shown here is generated by VoiceOver. I'm not sure how much control we have over it.

The focus of this ticket should be on the changes specified by design in this [UXPin](https://preview.uxpin.com/602167f0ca1beccf4bdf706e04b3a5bd34c78e94#/pages/139907964/documentation/no-panels?mode=i). I'd suggest we handle any unrelated major a11y-related changes in another ticket.

**Comment by nicko-amida** [comment]
@gescobardev Looks good for the current criteria! 

I'm also coming across a few more cases that I think can be taken into account, which I could compile for separate issue if needs be - for example:

* If the input field already has a valid search entry, upon regaining focus via keyboard tab or mouse click, the "Search for an available service" prompt doesn't need to show:
![image](https://user-images.githubusercontent.com/81988788/129382072-4dee56fb-f831-4c2a-90e5-3ec918c999d7.png)
* I guess this is more of a downshift.js thing, but spaces aren't ignored, so extra spaces discount an otherwise valid entry:
![image](https://user-images.githubusercontent.com/81988788/129382544-ca520b91-4bad-43dd-b0d3-a9ac4854cb5c.png)
* (I also came across the keyboard search thing @tlei123 mentioned above)

**Comment by gescobardev** [comment]
@tlei123 Thanks for catching this! Since this issue appears to be on prod, it seems reasonable to create a separate ticket for this, in that way I can tackle the other related bugs.
@joshkimux would you be able to open an A11y ticket for the above issue that @tlei123 found, when you get the chance?

**Comment by gescobardev** [comment]
@joshkimux Great catch with the dropdown list height. Since this seems like a quick fix, I can do that in this ticket. 
In regards to the instructions issue, I agree with @mpelzsherman. Not quite sure how much control we have over that. 
Could this be made into a separate issue as well? 
@AngelaFowler82 have you ran across this issue using VoiceOver elsewhere?

**Comment by gescobardev** [comment]
@nicko-amida Thanks for catching these! The issue with prompt showing when it shouldn't is well within scope and I can work on fixing that now. 
I can also look into the issue when a space is added after a service is selected. Though, I believe this issue is already happening on prod it just happens to be more noticeable now bc the error shows up. This may need to be a separate ticket but I'll keep you posted since I'm going to tackle your other comment first.

**Comment by joshkimux** [comment]
@gescobardev @tlei123 The error issue may be better logged as a bug rather than an accessibility defect as noted by @mpelzsherman earlier (it's been observed in prod already). I went ahead and made a ticket to help things get kicked off 😄 

https://github.com/department-of-veterans-affairs/va.gov-team/issues/26852#issuecomment-898158140

**Comment by gescobardev** [comment]
@joshkimux Perfect, thanks so much 😁

**Comment by mmiddaugh** [comment]
Spoke with @gescobardev this morning. This work on the service type field meets the AC but also revealed an usability issue which has been quietly happening on prod plus made it more evident. The usability issue could impact a broader audience so we'll work to resolve this new behavior (#28734) and ship the changes at the same time.

**Comment by gescobardev** [comment]
@joshkimux @tlei123 When you get the chance, can you re-review the RI?
It was decided to go ahead and fix the issue you noted with executing successful search on keyboard navigation within the same PR (separate ticket created by Josh - [#28734](https://github.com/department-of-veterans-affairs/va.gov-team/issues/28734))
Can you also confirm that when the `Search for an available service` or `We couldn't find that, please try another service` prompts display, that they are being announced to the user via VoiceOver?
Thank you (:

**Comment by joshkimux** [comment]
@gescobardev Just reviewed this now, I've attached feedback below!

### Longer list issue is fixed! 🥳 

<img width="1000" alt="Screen Shot 2021-08-24 at 9 33 19 AM" src="https://user-images.githubusercontent.com/14154792/130626153-b34f9c97-fc28-4a93-bbac-9af1fc71c446.png">

### Error messages `search for an available service` and `we couldn't find that...` do not reliably announce out to screen reader users

Errors are currently being announced exclusively using `role="alert"` which will announce errors once the first time, but not subsequently if the user edits the input to another string that draws the error. For example:

https://user-images.githubusercontent.com/14154792/130623111-0c65236a-d297-47c5-912d-4e2a27b042c0.mov

In this video, you'll notice that after the error announces for the first time (5s mark), it will fail to announce again after the user conducts a search (9s) and when they conduct a new incorrect search.

#### Recommendation

Consider using `aria-describedby` to link both errors to the input field itself. This will ensure that whenever the user focuses on the input field-- either naturally or when submitting a search-- the errors will be read out every time politely.

<img width="1792" alt="Screen Shot 2021-08-24 at 9 12 54 AM" src="https://user-images.githubusercontent.com/14154792/130623543-ee6ade70-3d96-4d9c-bb41-a2c65f9efd3a.png">

If possible, I'd also recommend tweaking the errors to be a bit more descriptive and unique so users can understand exactly what went wrong.


### `Role="alert"` can be very distracting to screen reader users

I believe that `role="alert"` is what we use for our error messages according to the [VA design system](https://design.va.gov/patterns/form-feedback)-- but in this particular use case it may be worth exploring using `role="status"` or `aria-live="polite"` for the error messaging instead because:
- `role="alert"` immediately announces the alert without waiting for the user to finish typing or pause; as a result, folks may type over the alert itself and miss it
- since the search combobox uses assertive messaging already either through `role="alert"` or `aria-live="assertive"`; the experience is already noisy and distracting with a number of "beeps" interrupting the flow of the user. Making error messaging more calm with polite messaging may be preferred.

#### Recommendation

To get this right, I'd recommend:
- testing error messaging using `aria-live="polite"` and `role="status"`
- working with @AngelaFowler82 on finding the right balance

**Comment by tlei123** [comment]
@gescobardev , I also just checked this on the RI, and my findings corroborate @joshkimux 's above.
